Convert transform types to strongly typed CSS/Style values

weinig opened the pull request:
#### 31e8573b41c815839a1e9f292594e7872414bbcb
<pre>
Convert transform types to strongly typed CSS/Style values
<a href="https://bugs.webkit.org/show_bug.cgi?id=282492">https://bugs.webkit.org/show_bug.cgi?id=282492</a>

Reviewed by NOBODY (OOPS!).

Replaces existing transform value implementation to have three
representations: CSS,  Style and Platform.

The CSS and Style representations model the transform functions
directly (like was done with shapes recently). In addition, types
for the CSS properties themselves (`transform`, `translate`,
`rotate` and `scale`) have value types along with `CSSValue`
subclasses to wrap them.

The Platform representation implements a reduced set of the transform
functions, removing the axis specific functions (e.g. `scaleX()`,
`rotateZ()`, etc), as they are not needed at that level. The Platform
representation is also &quot;absolute&quot;, meaning any percentages have been
resolved. This simplifies use in GraphicsLayer, as no reference size
needs to be passed in.

A naive implementation of the `&lt;transform-list&gt;` production would have
left been to use something like:

    `Vector&lt;std::variant&lt;MatrixFunction, Matrix3DFunction, ScaleFunction, ...&gt;&gt;`

This would be rather memory inefficient, as each element of the Vector
would be the size of the of the largest function, which in this case
would be Matrix3D at 128 bytes (+ tag size).

To avoid this, a new data structure, WTF::VariantList&lt;Ts...&gt; is introduced
which acts like a Vector&lt;std::variant&lt;Ts...&gt;&gt; except that each element only
takes up its own size (+ tag size). The tradeoff is that the list is not
random access capable, and access to the elements is only available via
iteration. In practice, this tradeoff is a good one for the CSS and Style
value types.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/VariantList.h: Added.
* Source/WTF/wtf/VariantListOperations.h: Added.
* Source/WTF/wtf/text/TextStream.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
* Source/WebCore/css/CSSBasicShapeValue.cpp:
* Source/WebCore/css/CSSBasicShapeValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSPathValue.cpp:
* Source/WebCore/css/CSSPathValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSRayValue.cpp:
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/CSSRotatePropertyValue.cpp: Added.
* Source/WebCore/css/CSSRotatePropertyValue.h: Added.
* Source/WebCore/css/CSSScalePropertyValue.cpp: Added.
* Source/WebCore/css/CSSScalePropertyValue.h: Added.
* Source/WebCore/css/CSSTransformFunctionValue.cpp: Added.
* Source/WebCore/css/CSSTransformFunctionValue.h: Added.
* Source/WebCore/css/CSSTransformPropertyValue.cpp: Added.
* Source/WebCore/css/CSSTransformPropertyValue.h: Added.
* Source/WebCore/css/CSSTranslatePropertyValue.cpp: Added.
* Source/WebCore/css/CSSTranslatePropertyValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueList.cpp:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/ComputedStyleDependencies.cpp: Added.
* Source/WebCore/css/ComputedStyleDependencies.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
* Source/WebCore/css/DeprecatedCSSOMValueList.h:
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/typedom/CSSKeywordValue.cpp:
* Source/WebCore/css/typedom/CSSKeywordValue.h:
* Source/WebCore/css/typedom/CSSNumericFactory.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.h:
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathValue.h:
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.h:
* Source/WebCore/css/typedom/transform/CSSPerspective.cpp:
* Source/WebCore/css/typedom/transform/CSSPerspective.h:
* Source/WebCore/css/typedom/transform/CSSRotate.cpp:
* Source/WebCore/css/typedom/transform/CSSRotate.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
* Source/WebCore/css/typedom/transform/CSSScale.h:
* Source/WebCore/css/typedom/transform/CSSSkew.cpp:
* Source/WebCore/css/typedom/transform/CSSSkew.h:
* Source/WebCore/css/typedom/transform/CSSSkewX.cpp:
* Source/WebCore/css/typedom/transform/CSSSkewX.h:
* Source/WebCore/css/typedom/transform/CSSSkewY.cpp:
* Source/WebCore/css/typedom/transform/CSSSkewY.h:
* Source/WebCore/css/typedom/transform/CSSTransformComponent.h:
* Source/WebCore/css/typedom/transform/CSSTransformValue.cpp:
* Source/WebCore/css/typedom/transform/CSSTransformValue.h:
* Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
* Source/WebCore/css/typedom/transform/CSSTranslate.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSOMConversion.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueVisitation.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.cpp: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.cpp:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/transforms/CSSRotateProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSRotateProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSScaleProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSScaleProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformFunctions.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTransformFunctions.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformList.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTransformProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSTranslateProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTranslateProperty.h: Added.
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerAnimationProperty.cpp: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationProperty.h: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationValue.cpp: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationValue.h: Added.
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.cpp:
* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformFunctionType.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformFunctionType.h: Renamed from Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.cpp.
* Source/WebCore/platform/graphics/transforms/TransformFunctions.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformFunctions.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformList.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformList.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.cpp: Renamed from Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp.
* Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.h: Renamed from Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h.
* Source/WebCore/platform/graphics/transforms/TransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperations.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.h: Removed.
* Source/WebCore/platform/mediastream/RealtimeOutgoingAudioSource.h:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/shapes/LayoutShape.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/TransformOperationsBuilder.cpp: Removed.
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.cpp:
* Source/WebCore/style/values/transforms/StyleIndividualTransformProperties.h: Added.
* Source/WebCore/style/values/transforms/StyleRotateProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleRotateProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleScaleProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleScaleProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformFunctions.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformFunctions.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformInterfaces.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformList.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformList.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleTranslateProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTranslateProperty.h: Added.
* Source/WebCore/svg/SVGGraphicsElement.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/VariantList.cpp: Added.
* Tools/TestWebKitAPI/Tests/WebCore/PlatformCAAnimationKeyPath.cpp:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31e8573b41c815839a1e9f292594e7872414bbcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77242 "240 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56277 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81810 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28524 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79359 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65425 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60512 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18555 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50478 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66292 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40812 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23790 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26847 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70427 "Found 15 jsc stress test failures: stress/get-by-val-hoist-above-structure-2.js.eager-jettison-no-cjit, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-b3o0, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-no-put-stack-validate, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-small-pool, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-validate-sampling-profiler ...") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68991 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83230 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76520 "Found 3 new JSC stress test failures: stress/attribute-custom-accessor.js.ftl-eager, stress/get-by-val-hoist-above-structure-2.js.eager-jettison-no-cjit, stress/get-by-val-hoist-above-structure.js.ftl-eager-no-cjit-b3o1 (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4622 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3117 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68044 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12030 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10127 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98773 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4569 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7384 "Found 7 new failures in css/CSSRotatePropertyValue.h, css/CSSScalePropertyValue.h, css/CSSTranslatePropertyValue.h, css/values/primitives/CSSPrimitiveNumericTypes.h, css/CSSTransformPropertyValue.h, css/CSSTransformFunctionValue.h and found 11 fixed files: css/typedom/transform/CSSRotate.cpp, css/typedom/transform/CSSTransformValue.cpp, platform/graphics/transforms/TransformOperations.cpp, platform/graphics/GraphicsLayer.h, css/CSSValue.cpp, platform/graphics/transforms/RotateTransformOperation.cpp, css/typedom/transform/CSSTranslate.cpp, css/typedom/transform/CSSPerspective.cpp, css/typedom/transform/CSSMatrixComponent.cpp, style/TransformOperationsBuilder.cpp ...") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21608 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4588 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8023 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6347 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.cpp:
```diff
@@ -24,37 +24,36 @@
-
-#include "TransformOperations.h"
-#include <algorithm>
+#include "TransformListSharedPrimitivesPrefix.h"
```
I think EWS says we also need to add `#include <array>` here, for Windows

darinadler commented on Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.cpp:
```diff
@@ -24,37 +24,36 @@
+        TransformFunctionType transform2D;
+        TransformFunctionType transform3D;
+    };
+    static constexpr std::array<Set, 3> sharedPrimitives = {
```
While looking for how we were using `std::array` in this file, I noticed that I think we can omit `<Set, 3>` here.

weinig committed:
Convert transform types to strongly typed CSS/Style values
https://bugs.webkit.org/show_bug.cgi?id=282492

Reviewed by NOBODY (OOPS!).

Replaces existing transform value implementation to have three
representations: CSS,  Style and Platform.

The CSS and Style representations model the transform functions
directly (like was done with shapes recently). In addition, types
for the CSS properties themselves (`transform`, `translate`,
`rotate` and `scale`) have value types along with `CSSValue`
subclasses to wrap them.

The Platform representation implements a reduced set of the transform
functions, removing the axis specific functions (e.g. `scaleX()`,
`rotateZ()`, etc), as they are not needed at that level. The Platform
representation is also "absolute", meaning any percentages have been
resolved. This simplifies use in GraphicsLayer, as no reference size
needs to be passed in.

A naive implementation of the `<transform-list>` production would have
left been to use something like:

    `Vector<std::variant<MatrixFunction, Matrix3DFunction, ScaleFunction, ...>>`

This would be rather memory inefficient, as each element of the Vector
would be the size of the of the largest function, which in this case
would be Matrix3D at 128 bytes (+ tag size).

To avoid this, a new data structure, WTF::VariantList<Ts...> is introduced
which acts like a Vector<std::variant<Ts...>> except that each element only
takes up its own size (+ tag size). The tradeoff is that the list is not
random access capable, and access to the elements is only available via
iteration. In practice, this tradeoff is a good one for the CSS and Style
value types.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/VariantList.h: Added.
* Source/WTF/wtf/VariantListOperations.h: Added.
* Source/WTF/wtf/text/TextStream.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
* Source/WebCore/css/CSSBasicShapeValue.cpp:
* Source/WebCore/css/CSSBasicShapeValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSPathValue.cpp:
* Source/WebCore/css/CSSPathValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSRayValue.cpp:
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/CSSRotatePropertyValue.cpp: Added.
* Source/WebCore/css/CSSRotatePropertyValue.h: Added.
* Source/WebCore/css/CSSScalePropertyValue.cpp: Added.
* Source/WebCore/css/CSSScalePropertyValue.h: Added.
* Source/WebCore/css/CSSTransformFunctionValue.cpp: Added.
* Source/WebCore/css/CSSTransformFunctionValue.h: Added.
* Source/WebCore/css/CSSTransformPropertyValue.cpp: Added.
* Source/WebCore/css/CSSTransformPropertyValue.h: Added.
* Source/WebCore/css/CSSTranslatePropertyValue.cpp: Added.
* Source/WebCore/css/CSSTranslatePropertyValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueList.cpp:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/ComputedStyleDependencies.cpp: Added.
* Source/WebCore/css/ComputedStyleDependencies.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
* Source/WebCore/css/DeprecatedCSSOMValueList.h:
* Source/WebCore/css/calc/CSSCalcTree+Copy.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/typedom/CSSKeywordValue.cpp:
* Source/WebCore/css/typedom/CSSKeywordValue.h:
* Source/WebCore/css/typedom/CSSNumericFactory.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.h:
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathValue.h:
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.cpp:
* Source/WebCore/css/typedom/transform/CSSMatrixComponent.h:
* Source/WebCore/css/typedom/transform/CSSPerspective.cpp:
* Source/WebCore/css/typedom/transform/CSSPerspective.h:
* Source/WebCore/css/typedom/transform/CSSRotate.cpp:
* Source/WebCore/css/typedom/transform/CSSRotate.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
* Source/WebCore/css/typedom/transform/CSSScale.h:
* Source/WebCore/css/typedom/transform/CSSSkew.cpp:
* Source/WebCore/css/typedom/transform/CSSSkew.h:
* Source/WebCore/css/typedom/transform/CSSSkewX.cpp:
* Source/WebCore/css/typedom/transform/CSSSkewX.h:
* Source/WebCore/css/typedom/transform/CSSSkewY.cpp:
* Source/WebCore/css/typedom/transform/CSSSkewY.h:
* Source/WebCore/css/typedom/transform/CSSTransformComponent.h:
* Source/WebCore/css/typedom/transform/CSSTransformValue.cpp:
* Source/WebCore/css/typedom/transform/CSSTransformValue.h:
* Source/WebCore/css/typedom/transform/CSSTranslate.cpp:
* Source/WebCore/css/typedom/transform/CSSTranslate.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSOMConversion.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueVisitation.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.cpp: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.cpp:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/transforms/CSSRotateProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSRotateProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSScaleProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSScaleProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformFunctions.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTransformFunctions.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformList.h: Added.
* Source/WebCore/css/values/transforms/CSSTransformProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTransformProperty.h: Added.
* Source/WebCore/css/values/transforms/CSSTranslateProperty.cpp: Added.
* Source/WebCore/css/values/transforms/CSSTranslateProperty.h: Added.
* Source/WebCore/dom/ViewTransition.cpp:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerAnimationProperty.cpp: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationProperty.h: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationValue.cpp: Added.
* Source/WebCore/platform/graphics/GraphicsLayerAnimationValue.h: Added.
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.cpp:
* Source/WebCore/platform/graphics/ca/PlatformCAAnimation.h:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperAnimation.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/Matrix3DTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/MatrixTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/PerspectiveTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/RotateTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/ScaleTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/SkewTransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformFunctionType.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformFunctionType.h: Renamed from Source/WebCore/platform/graphics/transforms/IdentityTransformOperation.cpp.
* Source/WebCore/platform/graphics/transforms/TransformFunctions.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformFunctions.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformList.cpp: Added.
* Source/WebCore/platform/graphics/transforms/TransformList.h: Added.
* Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.cpp: Renamed from Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.cpp.
* Source/WebCore/platform/graphics/transforms/TransformListSharedPrimitivesPrefix.h: Renamed from Source/WebCore/platform/graphics/transforms/TransformOperationsSharedPrimitivesPrefix.h.
* Source/WebCore/platform/graphics/transforms/TransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperation.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TransformOperations.h: Removed.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.cpp: Removed.
* Source/WebCore/platform/graphics/transforms/TranslateTransformOperation.h: Removed.
* Source/WebCore/platform/mediastream/RealtimeOutgoingAudioSource.h:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/shapes/LayoutShape.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/TransformOperationsBuilder.cpp: Removed.
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.cpp:
* Source/WebCore/style/values/transforms/StyleIndividualTransformProperties.h: Added.
* Source/WebCore/style/values/transforms/StyleRotateProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleRotateProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleScaleProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleScaleProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformFunctions.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformFunctions.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformInterfaces.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformList.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformList.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTransformProperty.h: Added.
* Source/WebCore/style/values/transforms/StyleTranslateProperty.cpp: Added.
* Source/WebCore/style/values/transforms/StyleTranslateProperty.h: Added.
* Source/WebCore/svg/SVGGraphicsElement.cpp:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/VariantList.cpp: Added.
* Tools/TestWebKitAPI/Tests/WebCore/PlatformCAAnimationKeyPath.cpp:
